### PR TITLE
mcp23xxx: support for Microchip MCP23 family of IO extenders

### PR DESCRIPTION
**Pull request by balazsgrill**
This driver supports the following I2C and SPI IO extenders:
* MCP23016
* MCP23017/MCP23S17
* MCP23018/MCP23S18
* MCP23008/MCP23S08
* MCP23009/MCP23S09

The driver is still under development, Interrupt support is not yet implemented.

### CODE REVIEW

**Comment by maruel** [comment]
Sorry for the delay. I can't approve because both the CLA checker and Travis failed. For the CLA checker, see the instructions above.

For travis, if you click through, you'll end up on https://travis-ci.org/google/periph/jobs/653761040 and scrolling a bit you'll find the red text:
```
Check Code is well formatted
The command "if [[ $TRAVIS_GO_VERSION != 1.7.6 ]]; then echo 'Check Code is well formatted'; ! gofmt -s -d . | read; fi" exited with 1.
```

Basically, run gofmt -s on the new files and it should do a small number of modifications.

**Comment by balazsgrill** [comment]
@googlebot I signed it!

**Comment by maruel** [comment]
gohci

**Comment by maruel** [comment]
Could you fix some of the comments? In practice the copyright one is the one I care the most about.

File: experimental/cmd/mcp23xxx/main.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package main
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/cmd/mcp23xxx/main.go]: done.

File: experimental/devices/mcp23xxx/doc.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package mcp23xxx controls IO extenders belonging to Microchip MCP23 family
+//
+// Supported variants:
+// * MCP23008/MCP23S08
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/doc.go]: For now, I removed this list from here and tried to document the supported variants at the appropriate constant declarations, although godoc doesn't render the comments there. Do you have any suggestions?

File: experimental/cmd/mcp23xxx/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/cmd/mcp23xxx/main.go]: done

File: experimental/devices/mcp23xxx/mcp23xxx.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import (
+	"fmt"
+	"strconv"
+
+	"periph.io/x/periph/conn/gpio/gpioreg"
+	"periph.io/x/periph/conn/i2c"
+	"periph.io/x/periph/conn/spi"
+)
+
+type Dev struct {
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/mcp23xxx.go]: done.

File: experimental/devices/mcp23xxx/pins.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/physic"
+	"periph.io/x/periph/conn/pin"
+)
+
+// MCP23xxxPin extends gpio.PinIO interface with features supported by MCP23xxx devices
+type MCP23xxxPin interface {
+	gpio.PinIO
+	// SetPolarity if set to Inverted, GPIO register bit reflects the same logic state of the input pin
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/pins.go]: done.

File: experimental/devices/mcp23xxx/pins.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+const (
+	Normal   Polarity = false
+	Inverted Polarity = true
+)
+
+type port struct {
+	name string
+
+	// GPIO basic registers
+	iodir *registerCache
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/pins.go]: done.

File: experimental/devices/mcp23xxx/registers.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+func (ra *spiRegisterAccess) writeRegister(address uint8, value uint8) error {
+	return ra.Tx([]byte{0x40, address, value}, nil)
+}
+
+func (ra *spiRegisterAccess) define(address uint8) *registerCache {
+	return newRegister(ra, address)
+}
+
+type registerCache struct {
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/registers.go]: Do you refer to the caching part, or how a register is accessed? I think this generalization between I2C and SPI register access (and having only 1-byte register values) is specific to this device family.

File: experimental/devices/mcp23xxx/mcp23xxx.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	Pins [][]MCP23xxxPin
+}
+
+type Variant string
+
+const (
+	MCP23x08 Variant = "MCP23%s08"
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/mcp23xxx.go]: I changed this part, having specific constants for each device.

File: experimental/devices/mcp23xxx/pins.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	"errors"
+	"strconv"
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/physic"
+	"periph.io/x/periph/conn/pin"
+)
+
+// MCP23xxxPin extends gpio.PinIO interface with features supported by MCP23xxx devices
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/pins.go]: done.

File: experimental/devices/mcp23xxx/mcp23xxx.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func makeDev(ra registerAccess, variant Variant, devicename string) (*Dev, error) {
+	var ports []*port
```
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/mcp23xxx.go]: done.

File: experimental/devices/mcp23xxx/pins.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/pins.go]: done.

File: experimental/devices/mcp23xxx/mcp23xxx.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (balazsgrill) [reply experimental/devices/mcp23xxx/mcp23xxx.go]: It took a while for me to understand what happens here in https://github.com/google/periph/pull/433/commits/68a60d80ba616bcd189dd7c004fc5c4c7392e8aa.
If I iterate through a slice of sructs by value then use a reference to it, it's actually a reference to the iterating variable which is overwritten on the next cycle. Should be fixed in https://github.com/google/periph/pull/433/commits/0c17bcc41ea9bfa3048de0580e5d0a2634a551ca.

File: experimental/devices/mcp23xxx/registers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (maruel) [reply experimental/devices/mcp23xxx/registers.go]: Ok

**Comment by maruel** [comment]
gohci

**Comment by maruel** [comment]
Thanks!

**Pull request merged**